org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceRule.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) OSGi Alliance (2019, 2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit4.service;
+
+import static org.osgi.test.common.inject.FieldInjector.findAnnotatedNonStaticFields;
+import static org.osgi.test.common.inject.FieldInjector.setField;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.junit.rules.MethodRule;
+import org.osgi.test.common.annotation.InjectService;
+import org.osgi.test.common.service.ServiceAware;
+import org.osgi.test.junit4.context.BundleContextRule;
+
+/**
+ * A JUnit 4 Rule to depend on OSGi services.
+ * <p>
+ * Example: <br>
+ *
+ * <pre>
+ * &#64;Rule
+ * public ServiceRule sr = new ServiceRule();
+ *
+ * &#64;InjectService
+ * Foo foo;
+ *
+ * &#64;Test
+ * public void aTest() {
+ * 	// use foo
+ * }
+ * </pre>
+ */
+public class ServiceRule extends BundleContextRule implements AutoCloseable, MethodRule {
```

Comment:
It seems odd to me that ServiceRule extends BundleContextRule. If that is necessary, then perhaps we only need one JUnit 4 rule?

Reply:
Unlike JUnit 5 Extensions which have the store,  there's no built-in mechanism in JUnit 4 rules to synchronize between rules. I'm guessing he's fine this to make the synchronization easier. But maybe composition would be better than inheritance?

Reply:
I moved it back to composition but you'll note that 99% of the time the default constructor will be used, anyway, it's not a big issue.